gossip/c_block_callbacks.go

```diff
@@ -247,14 +247,14 @@ func consensusCallbackBeginBlockFn(
 
 				// skip LLR block/epoch deciding if not activated
 				if !es.Rules.Upgrades.Llr {
-					llrs := store.GetLlrState()
-					if llrs.LowestBlockToDecide == blockCtx.Idx {
-						llrs.LowestBlockToDecide++
-					}
-					if sealing && es.Epoch+1 == llrs.LowestEpochToDecide {
-						llrs.LowestEpochToDecide++
-					}
-					store.SetLlrState(llrs)
+					store.ModifyLlrState(func(llrs *LlrState) {
```

Comment:
this block `if !es.Rules.Upgrades.Llr {` means node always apply the **Llr** upgrade then it will rollback the Llr state if the node config doesn't support Llr upgrade, is it correct?

Reply:
Almost - indeed it will always enable the Llr upgrade if so was commanded by `driverpos.Topics.UpdateNetworkRules` (given that the node was upgraded when it processed the tx with the upgrade and knew what Llr upgrade is, otherwise it will have to upgrade and resync from scratch).
This `if` is needed so that node would correctly maintain `LowestBlockToDecide` and `LowestEpochToDecide` for period(s) when Llr is disabled.

After each processed block, this piece of code will increment LowestBlockToDecide, effectively signaling that the just processed block doesn't need any more LLR votes (since block doesn't belong to LLR period, it doesn't need LLR votes at all). The same for epochs.

For example:
Let's assume that LLR gets enabled at block 70. After LLR is enabled, LLR will not look for votes for blocks 1-69
Without this code, LLR would start to work with LowestBlockToDecide=1 and would infinitely search votes for blocks 1-69 (which don't exist because LLR validators don't vote for blocks before LLR)